File: src/main/java/com/dyrosoft/kvbparser/parser/DeparturesParserFunc.java
package com.dyrosoft.kvbparser.parser;

import com.dyrosoft.kvbparser.models.Departure;
import com.dyrosoft.kvbparser.models.Line;
import com.dyrosoft.kvbparser.utils.StringUtils;
import com.google.common.collect.ImmutableList;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

public class DeparturesParserFunc extends AbstractRxHtmlParserFunc<ImmutableList<Departure>> {

    @Override
    protected ImmutableList<Departure> parse(final String html) {
        final ImmutableList.Builder<Departure> builder = ImmutableList.builder();

        final Document document = Jsoup.parse(html);
        final Elements elements = document.select("table.qr_table");
        for (final Element tr : elements.get(1).getElementsByTag("tr")) {
            final Elements tds = tr.getElementsByTag("td");
            final Departure departure =
                    new Departure(new Line(StringUtils.advacedTrim(tds.get(0).text())),
                                  StringUtils.advacedTrim(tds.get(1).text()),
                                  StringUtils.advacedTrim(tds.get(2).text()));
            builder.add(departure);
        }
        return builder.build();
    }
}

File: src/main/java/com/dyrosoft/kvbparser/parser/DeparturesParserFunc.kt
package com.dyrosoft.kvbparser.parser

import com.dyrosoft.kvbparser.models.Departure
import com.dyrosoft.kvbparser.models.Line
import com.dyrosoft.kvbparser.utils.StringUtils
import io.multifunctions.letNotNull
import org.jsoup.Jsoup

internal class DeparturesParserFunc : AbstractRxHtmlParserFunc<List<Departure>>() {

    override fun parse(html: String): List<Departure> {
        val list = mutableListOf<Departure>()
        val document = Jsoup.parse(html)
        val elements = document.select("table.qr_table")
        elements[1].getElementsByTag("tr")
                .map { it.getElementsByTag("td") }
                .forEach {
                    Triple(StringUtils.advancedTrim(it[0].text()),
                           StringUtils.advancedTrim(it[1].text()),
                           StringUtils.advancedTrim(it[2].text())).letNotNull { line, direction, waitTime ->

                        list.add(Departure(Line(line), direction, waitTime))
                    }
                }
        return list.toList()
    }
}
